vortex-duckdb/cpp/copy_function.cpp

```diff
@@ -0,0 +1,151 @@
+    }
+
+    auto c_column_types = vector<duckdb_logical_type>();
+    for (auto &col_type : column_types) {
+        c_column_types.push_back(reinterpret_cast<duckdb_logical_type>(const_cast<LogicalType *>(&col_type)));
+    }
+
+    duckdb_vx_error error_out = nullptr;
+    auto ffi_bind_data =
+        copy_vtab_one.bind(reinterpret_cast<duckdb_vx_copy_func_bind_input>(&info), c_column_names.data(),
+                           c_column_names.size(), c_column_types.data(), c_column_types.size(), &error_out);
+    if (error_out) {
+        throw BinderException(IntoErrString(error_out));
+    }
+
+    return make_uniq<CCopyBindData>(
+        // This should only be filled out once
+        copy_vtab_one, unique_ptr<CData>(reinterpret_cast<CData *>(ffi_bind_data)));
+}
+
+unique_ptr<GlobalFunctionData> c_init_global(ClientContext &context, FunctionData &bind_data,
+                                             const string &file_path) {
+    auto &bind = bind_data.Cast<CCopyBindData>();
+    duckdb_vx_error error_out = nullptr;
+    auto global_data = bind.vtab.init_global(bind.ffi_data->DataPtr(), file_path.c_str(), &error_out);
+    if (error_out) {
+        throw ExecutorException(IntoErrString(error_out));
+    }
+
+    return make_uniq<CCopyGlobalData>(unique_ptr<CData>(reinterpret_cast<CData *>(global_data)));
+}
+
+unique_ptr<LocalFunctionData> c_init_local(ExecutionContext &context, FunctionData &bind_data) {
+    auto &bind = bind_data.Cast<CCopyBindData>();
+    duckdb_vx_error error_out = nullptr;
+    auto data = bind.vtab.init_local(bind.ffi_data->DataPtr(), &error_out);
+    if (error_out) {
+        throw ExecutorException(IntoErrString(error_out));
+    }
+
+    return make_uniq<CCopyLocalData>(unique_ptr<CData>(reinterpret_cast<CData *>(data)));
+}
+
+void c_copy_to_sink(ExecutionContext &context, FunctionData &bind_data, GlobalFunctionData &gstate,
+                    LocalFunctionData &lstate, DataChunk &input) {
+    auto &bind = bind_data.Cast<CCopyBindData>();
+    auto &global = gstate.Cast<CCopyGlobalData>();
+    auto &local = lstate.Cast<CCopyLocalData>();
+    duckdb_vx_error error_out = nullptr;
+    bind.vtab.copy_to_sink(bind.ffi_data->DataPtr(), global.ffi_data->DataPtr(), local.ffi_data->DataPtr(),
+                           reinterpret_cast<duckdb_data_chunk>(&input), &error_out);
+    if (error_out) {
+        throw ExecutorException(IntoErrString(error_out));
+    }
+}
+
+void copy_to_finalize(ClientContext &context, FunctionData &bind_data, GlobalFunctionData &gstate) {
+    auto &bind = bind_data.Cast<CCopyBindData>();
+    auto &global = gstate.Cast<CCopyGlobalData>();
+    duckdb_vx_error error_out = nullptr;
+    bind.vtab.copy_to_finalize(bind.ffi_data->DataPtr(), global.ffi_data->DataPtr(), &error_out);
+    if (error_out) {
+        throw ExecutorException(IntoErrString(error_out));
+    }
+}
+
+extern "C" duckdb_vx_copy_func_vtab_t *get_vtab_one() {
```

Comment:
I guess strictly speaking we don't need a separate function, we can declare the global static var in the header and drop the get_vtab_one function.

Reply:
I prefer this fn